### PR TITLE
Upgrade kustomize and use the set image command to patch the image name and version

### PR DESCRIPTION
**Pull request by msau42**

### CODE REVIEW

File: deploy/kubernetes/overlays/dev/kustomization.yaml [hunk 1]
```diff
@@ -1,5 +1,22 @@
...
-- controller_images.yaml
-- node_images.yaml
+- controller_add_snapshotter.yaml
+- controller_always_pull.yaml
+- node_always_pull.yaml
+images:
+- name: REPLACEME/gcp-compute-persistent-disk-csi-driver
+  # Replace this with your private image name
+  newName: gcr.io/gke-release/gcp-compute-persistent-disk-csi-driver
+  newTag: "v0.4.0-gke.0"
```
Comment (verult) [inline deploy/kubernetes/overlays/dev/kustomization.yaml]: Should this be "latest"?

File: deploy/kubernetes/base/controller.yaml [hunk 2]
```diff
@@ -51,3 +51,4 @@ spec:
         - name: cloud-sa-volume
           secret:
             secretName: cloud-sa
+  volumeClaimTemplates: []
```
Comment (verult) [inline deploy/kubernetes/base/controller.yaml]: Why is this needed?

File: deploy/kubernetes/overlays/dev/kustomization.yaml [hunk 3]
```diff
@@ -1,5 +1,22 @@
...
+- name: REPLACEME/csi-provisioner
+  newName: quay.io/k8scsi/csi-provisioner
```
Comment (verult) [inline deploy/kubernetes/overlays/dev/kustomization.yaml]: Why do we use quay here but GCR elsewhere?

File: deploy/kubernetes/overlays/dev/kustomization.yaml (hunk 1 above)
Comment (msau42) [reply deploy/kubernetes/overlays/dev/kustomization.yaml]: We don't release any official "latest" tags, so I just put our latest supported image.  What I expect developers to do, is to replace both newName and newTag with their own repo and tags.

File: deploy/kubernetes/base/controller.yaml (hunk 2 above)
Comment (msau42) [reply deploy/kubernetes/base/controller.yaml]: It's a bug in kustomize, will add a comment: https://github.com/kubernetes-sigs/kustomize/issues/504

File: deploy/kubernetes/overlays/dev/kustomization.yaml (hunk 3 above)
Comment (msau42) [reply deploy/kubernetes/overlays/dev/kustomization.yaml]: We have a bit of a chicken and egg problem.  Currently in order to test a new deployment, we need to modify the staging deployment here, check it in, and then look at the staging test results.  

Ideally, I want to be able to validate a new deployment in a pull job first before needing to check it in.  So I want to use the dev overlay for that, but the pull jobs cannot access gke-staging images, so we need to use public images.

Comment (msau42) [reply deploy/kubernetes/overlays/dev/kustomization.yaml]: Maybe I'll rename this template to pull instead of dev to make it more obvious what it's purpose is for

File: deploy/kubernetes/overlays/prow-pull/kustomization.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+bases:
+- ../../base
+images:
+- name: REPLACEME/gcp-compute-persistent-disk-csi-driver
+  newName: gcr.io/gke-release/gcp-compute-persistent-disk-csi-driver
+  newTag: "v0.4.0-gke.0"
+- name: REPLACEME/csi-provisioner
+  newName: quay.io/k8scsi/csi-provisioner
```
Comment (davidz627) [inline deploy/kubernetes/overlays/prow-pull/kustomization.yaml]: the pull job should be using the gke release versions of sidecars right? and the `REPLACEME/gcp-compute-persistent-disk-csi-driver` should actually be replaced by a pull request specific built container?

File: deploy/kubernetes/overlays/dev/kustomization.yaml [hunk 5]
```diff
@@ -1,5 +1,22 @@
...
-- controller_images.yaml
-- node_images.yaml
+- controller_add_snapshotter.yaml
+- controller_always_pull.yaml
+- node_always_pull.yaml
+images:
+- name: REPLACEME/gcp-compute-persistent-disk-csi-driver
+  # Replace this with your private image names and tags
+  newName: gcr.io/gke-release/gcp-compute-persistent-disk-csi-driver
+  newTag: "v0.4.0-gke.0"
```
Comment (davidz627) [inline deploy/kubernetes/overlays/dev/kustomization.yaml]: should this instead be not replaced instead. Therefore when someone tries to use a dev overlay without changing anything they get an image pull failure (saying `REPLACEME` and reminding the user to replace to image) instead of failing silently by just giving them a working image.

File: deploy/kubernetes/overlays/prow-pull/kustomization.yaml (hunk 4 above)
Comment (msau42) [reply deploy/kubernetes/overlays/prow-pull/kustomization.yaml]: Ideally the pull job should use the gke versions of sidecars, however, we have a chicken and egg problem when we want to test a new version of a sidecar but it's not available in gke-release yet.  For example, [enabling topology](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/243).

The other option, is to not test new sidecar features in the pull job (only test stable features), and only switch over pull jobs when we change stable.  Actually, the pull job could just use the stable overlay.

File: deploy/kubernetes/overlays/dev/kustomization.yaml (hunk 5 above)
Comment (msau42) [reply deploy/kubernetes/overlays/dev/kustomization.yaml]: We currently use the dev overlay as our alpha features overlay too (for snapshot).  I can create a separate overlay for alpha if we want to separate that out

File: deploy/kubernetes/overlays/alpha/rbac_add_snapshots_to_provisioner.yaml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+# using jsonpatch
+# https://github.com/kubernetes-sigs/kustomize/blob/master/examples/jsonpatch.md
+- op: add
+  path: /rules/-
+  value:
+    apiGroups: ["snapshot.storage.k8s.io"]
+    resources: ["volumesnapshots"]
+    verbs: ["get", "list"]
+
+- op: add
```
Comment (davidz627) [inline deploy/kubernetes/overlays/alpha/rbac_add_snapshots_to_provisioner.yaml]: these have to be two seperate rules? We can't consolidate them somehow?

File: deploy/kubernetes/overlays/dev/kustomization.yaml [hunk 7]
```diff
@@ -1,5 +1,22 @@
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
 bases:
 - ../../base
 patches:
```
Comment (davidz627) [inline deploy/kubernetes/overlays/dev/kustomization.yaml]: shouldn't the dev overlay also have the `rbac_add_snapshotter.yaml`

File: deploy/kubernetes/overlays/prow-gke-head/kustomization.yaml [hunk 8]
```diff
@@ -1,5 +1,17 @@
...
+kind: Kustomization
 bases:
 - ../../base
-patches:
-- controller_images.yaml
-- node_images.yaml
+images:
+- name: REPLACEME/gcp-compute-persistent-disk-csi-driver
+  newName: gcr.io/gke-release-staging/gcp-compute-persistent-disk-csi-driver
+  newTag: "latest"
```
Comment (davidz627) [inline deploy/kubernetes/overlays/prow-gke-head/kustomization.yaml]: would it make more sense to rename this overlay `prow-gke-release-staging` or something like that?

Comment (davidz627) [reply deploy/kubernetes/overlays/prow-gke-head/kustomization.yaml]: wait nevermind, thats another overlay. `prow-gke-release-staging-latest`?

File: deploy/kubernetes/overlays/alpha/rbac_add_snapshots_to_provisioner.yaml (hunk 6 above)
Comment (msau42) [reply deploy/kubernetes/overlays/alpha/rbac_add_snapshots_to_provisioner.yaml]: Based on the syntax, I don't think it's possible to combine them.  The `value` field is not an array

File: deploy/kubernetes/overlays/dev/kustomization.yaml (hunk 7 above)
Comment (msau42) [reply deploy/kubernetes/overlays/dev/kustomization.yaml]: Rebased dev overlay on top of alpha overlay

File: deploy/kubernetes/overlays/prow-gke-head/kustomization.yaml (hunk 8 above)
Comment (msau42) [reply deploy/kubernetes/overlays/prow-gke-head/kustomization.yaml]: renamed all prow-gke overlays
